### PR TITLE
Feature/unit tests

### PR DESCRIPTION
**Pull request by laurencejackson**
Adds tests functions for Experiment class and adds Github actions for:

- linting
- testing
- reporting coverage to PR

### CODE REVIEW

File: mlops/Experiment.py [hunk 1]
```diff
@@ -70,59 +66,69 @@ def print_experiment_info(self):
...
 
-    def build_project_file(self):
+    def build_project_file(self, path: str = '.'):
         print('Building project file')
-        projectfile = ProjectFile(self.config)
+        projectfile = ProjectFile(self.config, path=path, use_localhost=self.use_localhost)
         projectfile.generate_yaml()
 
-    def run(self, remote: str = None, **kwargs):
-        num_retries = 1
```
Comment (hshuaib90) [inline mlops/Experiment.py:97]: Why do we no longer need this retry logic?

File: mlops/Experiment.py [hunk 2]
```diff
@@ -70,59 +66,69 @@ def print_experiment_info(self):
...
-                    # send instruction over SSH to run on remote location
-                    # todo
-                    pass
-                else:
-                    mlflow.run('.',
-                               experiment_id=self.experiment_id,
-                               use_conda=False,
-                               **kwargs)
-
-            except BuildError as error:
```
Comment (hshuaib90) [inline mlops/Experiment.py:123]: Why do we not need to catch these BuildErrors anymore?

File: mlops/ProjectFile.py [hunk 3]
```diff
@@ -3,15 +3,23 @@
...
+import os
 
 
-class ProjectFile:
+class ProjectFile():
 
-    def __init__(self, config, projectfile_name='MLProject'):
+    def __init__(self, config, path='.', projectfile_name='MLProject', use_localhost=False):
         self.config = config
         self.projectfile_name = projectfile_name
```
Comment (hshuaib90) [inline mlops/ProjectFile.py:13]: how is this different from config['project']['NAME']

File: mlops/ProjectFile.py [hunk 4]
```diff
@@ -3,15 +3,23 @@
...
+        self.project_path = os.path.join(path, projectfile_name)
+
+        if use_localhost:
+            volume_path = config['project']['LOCAL_VOLUME_MOUNT']
+        else:
+            volume_path = config['project']['VOLUME_MOUNT']
+
         self.project_dict = {'name': config['project']['NAME'].lower(),
-                             'docker_env': {'volumes': [config['project']['VOLUME_MOUNT']],
+                             'docker_env': {'volumes': [volume_path],
```
Comment (hshuaib90) [inline mlops/ProjectFile.py:22]: why does this need to be a list?

File: requirements.txt [hunk 5]
```diff
@@ -1,6 +1,6 @@
 mlflow
-dvc~=2.0.5
+# dvc~=2.0.5
```
Comment (hshuaib90) [inline requirements.txt:2]: can you just delete this? Are you sure DVC isn't a requirement?

File: tests/data/test_entry.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# Mock entry point
+
+def test():
+    # initialise network and datamodule
```
Comment (hshuaib90) [inline tests/data/test_entry.py:4]: whats the point of this?

**Review by hshuaib90 — CHANGES_REQUESTED** [review]
Can you remove the pyc pycache files please. Some comments left inline too.

- [ ] remove pycahce files

File: .dvc/.gitignore [hunk 7]
```diff
@@ -1,3 +0,0 @@
-/config.local
```
[the remark this answers is not in the document]
Comment (laurencejackson) [reply .dvc/.gitignore:1]: The .dvc integration into MLOps isn't at the point I want it to be so I've been removing mentions of it for now. We should still use DVC but it's something that will need to be worked on as a future issue. 

Users can still use DVC manually using the CLI, and I think that's the approach we should take until we work out how we want to the integration to work.

File: .github/workflows/python-app.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      - '*'
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    env:
+      MYSQL_DATABASE: mlflow_db
+      MYSQL_USER: mlflow
+      MYSQL_PASSWORD: strongpassword1
```
[the remark this answers is not in the document]
Comment (laurencejackson) [reply .github/workflows/python-app.yml:21]: These are just examples variables for the test environment. I will move these to github secrets anyway for best practice.

File: mlflow_server/docker-compose.yml [hunk 9]
```diff
@@ -39,7 +39,7 @@ services:
         image: mlflow_nginx
         container_name: mlflow_nginx
         ports:
-            - "80:80"
+            - "85:85"
```
[the remark this answers is not in the document]
Comment (laurencejackson) [reply mlflow_server/docker-compose.yml:42]: Port 80 is used by XNAT, moved to avoid conflict.

File: mlops/ProjectFile.py (hunk 3 above)
Comment (laurencejackson) [reply mlops/ProjectFile.py:13]: This refers to the project file which has the filename MLProject - this is different to the project name. The MLProject file tells mlflow how to configure the docker environment when running a project

File: mlops/ProjectFile.py (hunk 4 above)
Comment (laurencejackson) [reply mlops/ProjectFile.py:22]: Not entirely sure of the reason, but there's an exception to catch it if its not a list!

`mlflow.exceptions.ExecutionException: Project configuration (MLproject file) was invalid: Docker volumes must be a list of strings, e.g.: '["/path1/:/path1", "/path2/:/path2"])`

**Review by hshuaib90 — APPROVED** [review]

File: mlops/Experiment.py (hunk 1 above)
Comment (laurencejackson) [reply mlops/Experiment.py:97]: Same as comment below - improved the error handling method here

**Pull request merged**